### PR TITLE
Introduce client_credentials flow support with GenericAuthority

### PR DESCRIPTION
**Pull request by g7ed6e**
Relates to #1538 

**Changes proposed in this request**
Introduce a new `GenericAuthority` type.
Introduce 2 public API to `AbstractApplicationBuilder<T>`:
- `T WithGenericAuthority(string authorityUri, string discoveryDocumentAddress = null)`
- `T WithDocumentRetriever(IDocumentRetriever documentRetriever)`

**Testing**
[x] End to end tests of a client_credentials flow assert an access_token is retrieved.
[] End to end tests of a user flow should assert an exception is thrown.

**Performance impact**

**Documentation**
- [ ] All relevant documentation is updated.

### CODE REVIEW

**Comment by bgavrilMS** [comment]
Hi @g7ed6e - thanks for the contribution, we will take a look! 
I invited you to a contributor to the repo, so that you don't have to mess around with forks. Both ways work, but it's a bit easier to manage contributions from inside.

File: src/client/Microsoft.Identity.Client/AppConfig/AbstractApplicationBuilder.cs [hunk 1]
```diff
@@ -894,6 +894,25 @@ public T WithB2CAuthority(string authorityUri)
...
         }
 
+        public T WithGenericAuthority(string authorityUri, string discoveryDocumentAddress = null)
+        {
+            var authorityInfo = AuthorityInfo.FromGenericAuthority(authorityUri, discoveryDocumentAddress);
+            Config.Authority = GenericAuthority.CreateAuthority(authorityInfo);
+            return (T)this;
+        }
+        
+        public T WithDocumentRetriever(IDocumentRetriever documentRetriever)
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/AppConfig/AbstractApplicationBuilder.cs:904]: Are you missing some files? This `IDocumentRetriever` seems to be undefined.

Comment (g7ed6e) [reply src/client/Microsoft.Identity.Client/AppConfig/AbstractApplicationBuilder.cs:904]: `IDocumentRetriever` comes from Microsoft.IdentityModel.Protocols nuget package. I took a dependency on Microsoft.IdentityModel.Protocols.OpenIdConnect in Microsoft.IdentityModel.Client.

**Comment by g7ed6e** [comment]
> @g7ed6e please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree

**Comment by g7ed6e** [comment]
> Hi @g7ed6e - thanks for the contribution, we will take a look! 
> 
> I invited you to a contributor to the repo, so that you don't have to mess around with forks. Both ways work, but it's a bit easier to manage contributions from inside.

Thanks @bgavrilMS !

**Comment by g7ed6e** [comment]
Closed in favor of #3945 which track the correct branch

**Pull request closed without merging**